File: hiob_contracts/ares_v3_make_context_v1.py
"""Atomic, server-owned make context for one Ares V3 command."""

from __future__ import annotations

from typing import Any, Literal, Mapping

from pydantic import BaseModel, Field, model_validator

from .ares_script_revision_v1 import (
    DigestStr,
    NonBlankStr,
    UuidStr,
    _FROZEN_STRICT,
)
from .brand_scope import CanonicalBrandSlug, canonical_brand_slug
from .factory import sha256_digest


_DIGEST_FIELDS = (
    "workspace_id",
    "run_id",
    "brand_slug",
    "subject_id",
    "product_id",
    "character_lock_digest",
    "character_lock_version",
    "product_lock_digest",
    "artemis_approval_receipt_id",
    "artemis_approval_receipt_digest",
    "artemis_approval_state_revision",
)


def derive_ares_v3_make_context_digest_v1(
    value: Mapping[str, Any] | BaseModel,
) -> str:
    """Digest the ready snapshot without command execution identifiers."""

    data = (
        value.model_dump(mode="json")
        if isinstance(value, BaseModel)
        else dict(value)
    )
    payload = {field: data[field] for field in _DIGEST_FIELDS}
    payload["brand_slug"] = canonical_brand_slug(payload["brand_slug"])
    return sha256_digest(payload)


class AresV3MakeContextV1(BaseModel):
    """The only public make-readiness contract used by Star."""

    model_config = _FROZEN_STRICT

    contract_version: Literal["AresV3MakeContext.v1"]
    workspace_id: UuidStr
    run_id: UuidStr
    brand_slug: CanonicalBrandSlug
    subject_id: NonBlankStr
    product_id: NonBlankStr
    character_lock_digest: DigestStr
    character_lock_version: int = Field(
        gt=0,
        le=9_007_199_254_740_991,
        strict=True,
    )
    product_lock_digest: DigestStr
    artemis_approval_receipt_id: NonBlankStr
    artemis_approval_receipt_digest: DigestStr
    artemis_approval_state_revision: int = Field(
        gt=0,
        le=9_007_199_254_740_991,
        strict=True,
    )
    make_context_digest: DigestStr

    @model_validator(mode="after")
    def _bind_make_context(self) -> "AresV3MakeContextV1":
        if (
            self.make_context_digest
            != derive_ares_v3_make_context_digest_v1(self)
        ):
            raise ValueError(
                "make_context_digest does not match server make context"
            )
        return self


__all__ = [
    "AresV3MakeContextV1",
    "derive_ares_v3_make_context_digest_v1",
]

File: ts/src/ares-v3-make-context-v1.ts
/** Atomic, server-owned make context for one Ares V3 command. */
import { z } from 'zod';

import { sha256Digest } from './factory/digest.js';

const NonBlankString = z.string().refine(
  (value) => value.trim().length > 0,
  'string must not be blank',
);
const DigestSchema = z
  .string()
  .regex(/^sha256:[0-9a-f]{64}$/, 'digest must be sha256:<64 lowercase hex>');
const UuidSchema = z
  .string()
  .regex(
    /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/,
    'UUID must use canonical lowercase form',
  );

function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === 'object') {
    for (const child of Object.values(value as Record<string, unknown>)) {
      deepFreeze(child);
    }
    Object.freeze(value);
  }
  return value;
}

export function deriveAresV3MakeContextDigestV1(
  value: Record<string, unknown>,
): string {
  const field = (name: string): unknown => {
    if (!(name in value) || value[name] === undefined) {
      throw new TypeError(`${name} is required for make context digest`);
    }
    return value[name];
  };
  return sha256Digest({
    workspace_id: field('workspace_id'),
    run_id: field('run_id'),
    brand_id: field('brand_id'),
    subject_id: field('subject_id'),
    product_id: field('product_id'),
    character_lock_digest: field('character_lock_digest'),
    character_lock_version: field('character_lock_version'),
    product_lock_digest: field('product_lock_digest'),
    artemis_approval_receipt_id: field('artemis_approval_receipt_id'),
    artemis_approval_receipt_digest:
      field('artemis_approval_receipt_digest'),
    artemis_approval_state_revision:
      field('artemis_approval_state_revision'),
  });
}

export const AresV3MakeContextV1Schema = z
  .object({
    contract_version: z.literal('AresV3MakeContext.v1'),
    workspace_id: UuidSchema,
    run_id: UuidSchema,
    brand_id: UuidSchema,
    subject_id: NonBlankString,
    product_id: NonBlankString,
    character_lock_digest: DigestSchema,
    character_lock_version: z.number().int().safe().positive(),
    product_lock_digest: DigestSchema,
    artemis_approval_receipt_id: NonBlankString,
    artemis_approval_receipt_digest: DigestSchema,
    artemis_approval_state_revision: z.number().int().safe().positive(),
    make_context_digest: DigestSchema,
  })
  .strict()
  .superRefine((value, ctx) => {
    if (
      value.make_context_digest
      !== deriveAresV3MakeContextDigestV1(value)
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: 'make_context_digest does not match server make context',
        path: ['make_context_digest'],
      });
    }
  })
  .transform(deepFreeze);

export type AresV3MakeContextV1 = z.infer<
  typeof AresV3MakeContextV1Schema
>;
